File: sdk/dart/test/e2e_test.dart
// Copyright 2026 Oddbit (https://oddbit.id)
// SPDX-License-Identifier: Apache-2.0
//
// End-to-end tests run by scripts/test-sdks-e2e.sh against a live
// `wrangler dev` instance. Requires:
//   SHRTNR_TEST_URL        base URL of the running dev server
//   SHRTNR_TEST_API_KEY    a create+read API key minted by the harness
//
// Skipped from the default test run via the `e2e` tag. Invoke with
// `dart test --tags e2e`.

@Tags(['e2e'])
library;

import 'dart:io';

import 'package:shrtnr/shrtnr.dart';
import 'package:test/test.dart';

void main() {
  final baseUrl = Platform.environment['SHRTNR_TEST_URL'];
  final apiKey = Platform.environment['SHRTNR_TEST_API_KEY'];

  if (baseUrl == null || apiKey == null) {
    // Fail hard rather than skip. This file is excluded from default
    // `dart test` via --exclude-tags e2e; when it runs we are inside the
    // scripts/test-sdks-e2e.sh harness, which must export both env vars.
    // A silent skip would hide a broken harness behind a green CI check.
    test('e2e env vars must be set', () {
      fail(
        'SHRTNR_TEST_URL and SHRTNR_TEST_API_KEY must be set. '
        'Run e2e tests via scripts/test-sdks-e2e.sh from the repo root, not directly.',
      );
    });
    return;
  }

  final client = ShrtnrClient(baseUrl: baseUrl, apiKey: apiKey);

  group('Dart SDK e2e — live wrangler dev', () {
    test('link lifecycle — create, get, delete', () async {
      final link = await client.links.create(
        url: 'https://example.com/dart-e2e',
        label: 'dart-e2e',
      );
      expect(link.url, 'https://example.com/dart-e2e');
      final fetched = await client.links.get(link.id);
      expect(fetched.id, link.id);
      final result = await client.links.delete(link.id);
      expect(result.deleted, isTrue);
    });

    test('slug mutations work against public routes', () async {
      final link = await client.links.create(
        url: 'https://example.com/dart-slugs',
      );
      try {
        await client.slugs.add(link.id, 'dart-e2e-slug');
        final disabled = await client.slugs.disable(link.id, 'dart-e2e-slug');
        expect(disabled.disabledAt, isNotNull);
        final enabled = await client.slugs.enable(link.id, 'dart-e2e-slug');
        expect(enabled.disabledAt, isNull);
        final removed = await client.slugs.remove(link.id, 'dart-e2e-slug');
        expect(removed.removed, isTrue);
      } finally {
        await client.links.delete(link.id);
      }
    });
  });
}

File: sdk/typescript/tests/e2e/e2e.test.ts
// Copyright 2026 Oddbit (https://oddbit.id)
// SPDX-License-Identifier: Apache-2.0
//
// End-to-end tests run by scripts/test-sdks-e2e.sh against a live
// `wrangler dev` instance. Requires:
//   SHRTNR_TEST_URL        base URL of the running dev server
//   SHRTNR_TEST_API_KEY    a create+read API key minted by the harness
// Default `yarn test` excludes this path via vitest.config.ts, so this
// suite only runs via the dedicated vitest.e2e.config.ts used by the
// harness. If either env var is missing at runtime, beforeAll throws so
// a misconfigured harness can't pass silently.

import { describe, it, expect, beforeAll } from "vitest";
import { ShrtnrClient } from "../../src";

const BASE_URL = process.env.SHRTNR_TEST_URL;
const API_KEY = process.env.SHRTNR_TEST_API_KEY;

describe("TS SDK e2e — live wrangler dev", () => {
  let client: ShrtnrClient;

  beforeAll(() => {
    // Fail hard rather than skip. These tests only run via the explicit
    // tests/e2e/ path (default `yarn test` excludes this folder), so missing
    // env vars at this point means the harness is misconfigured. A silent
    // skip would hide it behind a green CI check.
    if (!BASE_URL || !API_KEY) {
      throw new Error(
        "SHRTNR_TEST_URL and SHRTNR_TEST_API_KEY must be set. " +
          "Run e2e tests via scripts/test-sdks-e2e.sh from the repo root, not directly.",
      );
    }
    client = new ShrtnrClient({ baseUrl: BASE_URL, apiKey: API_KEY });
  });

  it("link lifecycle — create, get, delete", async () => {
    const link = await client.links.create({
      url: "https://example.com/ts-e2e",
      label: "ts-e2e",
    });
    expect(link.url).toBe("https://example.com/ts-e2e");
    const fetched = await client.links.get(link.id);
    expect(fetched.id).toBe(link.id);
    const del = await client.links.delete(link.id);
    expect(del.deleted).toBe(true);
  });

  it("slug mutations work against live routes", async () => {
    const link = await client.links.create({ url: "https://example.com/ts-slugs" });
    await client.slugs.add(link.id, "ts-e2e-slug");
    const disabled = await client.slugs.disable(link.id, "ts-e2e-slug");
    expect(disabled.disabledAt).not.toBeNull();
    const enabled = await client.slugs.enable(link.id, "ts-e2e-slug");
    expect(enabled.disabledAt).toBeNull();
    const removed = await client.slugs.remove(link.id, "ts-e2e-slug");
    expect(removed.removed).toBe(true);
    await client.links.delete(link.id);
  });
});
